web/src/people.ts

```diff
@@ -1951,6 +1972,95 @@ export function populate_valid_user_ids(params: StateData["user_groups"]): void
     }
 }
 
+function get_unique_key_for_user_fetch(user_ids: Set<number>): string {
+    const fetch_unique_key = [...user_ids].sort().join(",");
+    return fetch_unique_key;
+}
+
+function start_fetch_for_requested_users(): void {
+    const user_ids_to_fetch = fetch_users_storage.pending_user_ids;
+    fetch_users_storage.pending_user_ids = new Set();
+    fetch_users_storage.in_transit_user_ids =
+        fetch_users_storage.in_transit_user_ids.union(user_ids_to_fetch);
+    const fetch_unique_key = get_unique_key_for_user_fetch(user_ids_to_fetch);
+    fetch_users_storage.promise_for_in_transit.set(fetch_unique_key, {
+        promise: fetch_users_storage.promise_for_pending!,
+        resolver: fetch_users_storage.promise_resolver_for_pending!,
+    });
+    fetch_users_storage.promise_for_pending = undefined;
+    fetch_users_storage.promise_resolver_for_pending = undefined;
+
+    void fetch_users(user_ids_to_fetch).then((fetched_users) => {
+        for (const user of fetched_users) {
+            if (user.is_active) {
+                add_active_user(user);
+            } else {
+                non_active_user_dict.set(user.user_id, user);
+                _add_user(user);
+            }
+        }
+
+        // Resolve promises waiting on this fetch after updating the data locally.
+        fetch_users_storage.promise_for_in_transit.get(fetch_unique_key)!.resolver();
+        // Clean up in transit promise for this fetch.
+        fetch_users_storage.promise_for_in_transit.delete(fetch_unique_key);
+        // Remove fetched users from in transit user ids.
+        fetch_users_storage.in_transit_user_ids =
+            fetch_users_storage.in_transit_user_ids.difference(user_ids_to_fetch);
+    });
+}
+
+export async function fetch_users_from_ids(user_ids: number[]): Promise<void> {
+    // NOTE: NEVER USE THIS FUNCTION DIRECTLY.
+    // Call get_or_fetch_users_from_ids instead.
+    const unknown_user_ids = new Set(
+        user_ids.filter((user_id) => !people_by_user_id_dict.has(user_id)),
+    );
+
+    // We already have data for all requested users.
+    if (unknown_user_ids.size === 0) {
+        return;
+    }
+
+    // If we have a fetch in progress which contains the unknown user_ids,
+    // return a promise that resolves when that fetch completes.
+    if (
+        unknown_user_ids.intersection(fetch_users_storage.in_transit_user_ids).size ===
+        unknown_user_ids.size
+    ) {
+        const key = get_unique_key_for_user_fetch(fetch_users_storage.in_transit_user_ids);
+        return fetch_users_storage.promise_for_in_transit.get(key)!.promise;
```

Comment:
I don't think this logic works if multiple requests end up in flight. The key is a function of `fetch_users_storage.in_transit_user_ids`, but that value gets mutated to the union of the ongoing fetches and the latest fetch when we start a second fetch with one already in flight.

Reply:
Nice catch! Added `get_combined_promise_for_user_ids` for this.